api/prometheus/v1/remote/remote_api.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package remote implements bindings for Prometheus Remote APIs.
+package remote
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/klauspost/compress/snappy"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/prometheus/client_golang/api"
+	"github.com/prometheus/client_golang/internal/github.com/efficientgo/core/backoff"
+)
+
+// API is a client for Prometheus Remote Protocols.
+// NOTE(bwplotka): Only https://prometheus.io/docs/specs/remote_write_spec_2_0/ is currently implemented,
+// read protocols to be implemented if there will be a demand.
+type API struct {
+	client api.Client
+	opts   apiOpts
+
+	reqBuf, comprBuf []byte
+}
+
+// APIOption represents a remote API option.
+type APIOption func(o *apiOpts) error
+
+// TODO(bwplotka): Add "too old sample" handling one day.
+type apiOpts struct {
+	logger           *slog.Logger
+	backoff          backoff.Config
+	compression      Compression
+	retryOnRateLimit bool
+}
+
+var defaultAPIOpts = &apiOpts{
+	backoff: backoff.Config{
+		Min:        1 * time.Second,
+		Max:        10 * time.Second,
+		MaxRetries: 10,
+	},
+	// Hardcoded for now.
+	retryOnRateLimit: true,
+	compression:      SnappyBlockCompression,
+}
+
+// WithAPILogger returns APIOption that allows providing slog logger.
+// By default, nothing is logged.
+func WithAPILogger(logger *slog.Logger) APIOption {
+	return func(o *apiOpts) error {
+		o.logger = logger
+		return nil
+	}
+}
+
+type nopSlogHandler struct{}
+
+func (n nopSlogHandler) Enabled(context.Context, slog.Level) bool  { return false }
+func (n nopSlogHandler) Handle(context.Context, slog.Record) error { return nil }
+func (n nopSlogHandler) WithAttrs([]slog.Attr) slog.Handler        { return n }
+func (n nopSlogHandler) WithGroup(string) slog.Handler             { return n }
+
+// NewAPI returns a new API for the clients of Remote Write Protocol.
+//
+// It is not safe to use the returned API from multiple goroutines, create a
+// separate *API for each goroutine.
+func NewAPI(c api.Client, opts ...APIOption) (*API, error) {
```

Comment:
We could consider removing error from here and from `APIOption` as we don't have any option with validation or so. However I added it for future compatibility (we won't be able to easily extend it without breaking compatibility). Note: Currently api/ is experimenta... 🙈  I think I will remove error here and hope for the best, YAGNI.

```suggestion
func NewAPI(c api.Client, opts ...APIOption) *API {
```